cmd/erasure-readfile.go

```diff
@@ -44,117 +113,120 @@ func (s ErasureStorage) ReadFile(writer io.Writer, volume, path string, offset,
+	//
+	// The start index of the erasure coded block containing the
+	// `offset` byte of data is:
+	partDataStartIndex := (offset / blocksize) * chunksize
+	// The start index of the erasure coded block containing the
+	// (last) byte of data at the index `offset + length - 1` is:
+	blockStartIndex := ((offset + length - 1) / blocksize) * chunksize
+	// However, we need the end index of the e.c. block containing
+	// the last byte - we need to check if that block is the last
+	// block in the part (in that case, it may be have a different
+	// chunk size)
+	isLastBlock := (totalLength-1)/blocksize == (offset+length-1)/blocksize
+	var partDataEndIndex int64
+	if isLastBlock {
+		lastBlockChunkSize := chunksize
+		if totalLength%blocksize != 0 {
+			lastBlockChunkSize = ceilFrac(totalLength%blocksize, int64(s.dataBlocks))
+		}
+		partDataEndIndex = blockStartIndex + lastBlockChunkSize - 1
+	} else {
+		partDataEndIndex = blockStartIndex + chunksize - 1
+	}
+
+	// Thus, the length of data to be read from the part file(s) is:
+	partDataLength := partDataEndIndex - partDataStartIndex + 1
+	// The calculation above does not apply when length == 0:
+	if length == 0 {
+		partDataLength = 0
+	}
+
+	var buffers [][]byte
+	buffers, err = s.readConcurrent(volume, path, partDataStartIndex, partDataLength, verifiers)
+	if err != nil {
+		// Could not read enough disks.
+		return
 	}
-	lastBlock := totalLength / blocksize
-	startOffset := offset % blocksize
-	chunksize := getChunkSize(blocksize, s.dataBlocks)
 
 	blocks := make([][]byte, len(s.disks))
+	needsReconstruction := false
 	for i := range blocks {
-		blocks[i] = make([]byte, chunksize)
+		if buffers[i] != nil {
+			blocks[i] = make([]byte, chunksize)
+		} else {
+			blocks[i] = make([]byte, 0, chunksize)
+			if i < s.dataBlocks {
+				needsReconstruction = true
+			}
+		}
 	}
-	for off := offset / blocksize; length > 0; off++ {
-		blockOffset := off * chunksize
 
-		if currentBlock := (offset + f.Size) / blocksize; currentBlock == lastBlock {
-			blocksize = totalLength % blocksize
-			chunksize = getChunkSize(blocksize, s.dataBlocks)
+	numChunks := ceilFrac(partDataLength, chunksize)
+	copyLen := chunksize
+	for chunkNumber := int64(0); chunkNumber < numChunks; chunkNumber++ {
+		if chunkNumber == numChunks-1 {
+			if partDataLength%chunksize != 0 {
+				copyLen = partDataLength % chunksize
+			}
+			for i := range blocks {
+				if len(buffers[i]) != 0 {
+					blocks[i] = make([]byte, copyLen)
+				} else {
+					blocks[i] = make([]byte, 0, copyLen)
+				}
+			}
+		} else {
 			for i := range blocks {
-				blocks[i] = blocks[i][:chunksize]
+				if len(buffers[i]) == 0 {
+					blocks[i] = blocks[i][0:0]
+				}
 			}
 		}
-		err = s.readConcurrent(volume, path, blockOffset, blocks, verifiers, errChans)
-		if err != nil {
-			return f, errors.Trace(errXLReadQuorum)
+
+		for i := range blocks {
+			if len(buffers[i]) != 0 {
+				_ = copy(blocks[i], buffers[i][chunkNumber*chunksize:])
```

Comment:
Is there a way to avoid copying here, like `blocks[i] = buffers[i][chunkNumber*chunkSize:]` ? I guess to gain some performance.

Reply:
We also don't need to do `_ =` to avoid the return.

Reply:
Also we need to copy here @vadmeste

Reply:
Ah ok, I will try.

Reply:
@harshavardhana It looks like we can avoid copying.

Reply:
@vadmeste @harshavardhana I've added this change to reduce copying in a separate commit to aid in reviewing - you can see the change more clearly I think. I can squash it too if thats preferable.